lib/src/fortio.cpp

```diff
@@ -0,0 +1,384 @@
+ * not return a value. Instead, an exception is thrown, and all function use
+ * function-try-blocks. While it's generally bad practice to throw from
+ * destructors, the eclfio functions are extern C'd, and no exceptions leak.
+ * Since there are no other destructors run from these functions than that of
+ * fguard, the destructor-exception is ok, since it never happens during stack
+ * unwinding.
+ *
+ * A string-less exception type is used, and std::exception() is noexcept in
+ * C++11.
+ */
+struct fguard {
+    FILE* fp;
+    std::fpos_t pos;
+
+    fguard( FILE* x ) : fp( x ) {
+        const auto err = std::fgetpos( fp, &this->pos );
+
+        if( err ) {
+            /*
+             * when fgetpos fails, presumably everything is broken and we
+             * didn't even record our previous position. no point even trying
+             * to roll back from this, so set fp to nullptr and bail
+             */
+            this->fp = nullptr;
+            throw std::exception();
+        }
+    }
+
+    ~fguard() noexcept( false ) {
+        if( !this->fp ) return;
+
+        const auto err = std::fsetpos( fp, &this->pos );
+        if( err ) throw std::exception();
+    }
+};
+
+struct options {
+    bool bigendian  = true;
+    bool size_limit = true;
+    int elemsize    = sizeof( std::int32_t );
+
+    // transform by default unless record is nullptr and this record should be
+    // skipped
+    bool transform    = true;
+    bool force_notail = false;
+    bool allow_notail = false;
+};
+
+options parse_opts( const char* opts ) noexcept {
+    options o;
+
+    for( const char* op = opts; *op != '\0'; ++op ) {
+        switch( *op ) {
+            case 'c': o.elemsize = sizeof( char );         break;
+            case 'i': o.elemsize = sizeof( std::int32_t ); break;
+            case 'f': o.elemsize = sizeof( float );        break;
+            case 'd': o.elemsize = sizeof( double );       break;
+
+            case 'E': o.bigendian = true;                  break;
+            case 'e': o.bigendian = false;                 break;
+
+            case 't': o.transform = true;                  break;
+            case 'T': o.transform = false;                 break;
+
+            case '~': o.force_notail = true;               break;
+            case '$': o.allow_notail = true;               break;
+
+            case '#': o.size_limit = false;                break;
+            default: break;
+        }
+    }
+
+    return o;
+}
+
+/*
+ * to/from host carefully swaps endianness for a large array.
+ *
+ * The switch guarantees that byte swap happens properly, even on platforms
+ * where sizeof(char) == sizeof(int32), without accidently terminating before,
+ * relying on htons being well-defined on such platforms. Should any non-char,
+ * int16, and int32 sizes be passed through, nothing should happen, but
+ * non-zero is returned.
+ */
+
+int to_host( void* dst, int size, std::int32_t nmemb ) {
```

Comment:
When adding these functions can you at the same the remove the endian swap functions from `util.c`

Reply:
I think we can afford to wait a bit longer, it's they're still (somewhat) in use.

```
ecl/ecl_kw.cpp:276:    util_endian_flip_vector(buffer, sizeof_iotype, ecl_kw->size);
ecl/ecl_kw.cpp:296:  util_endian_flip_vector(buffer, sizeof_iotype, ecl_kw->size);
ecl/ecl_kw.cpp:315:      util_endian_flip_vector(buffer, sizeof_iotype, ecl_kw->size);
ecl/ecl_kw.cpp:1277:        util_endian_flip_vector(io_buffer, sizeof_iotype, int_vector_size(index_map));
ecl/ecl_kw.cpp:1388:      util_endian_flip_vector(&size , sizeof size , 1);
ecl/ecl_kw.cpp:1728:      util_endian_flip_vector(&size , sizeof size , 1);

lib/rms/rms_tagkey.cpp:458:      util_endian_flip_vector(tagkey->data , tagkey->sizeof_ctype , tagkey->size);
```

Reply:
OK - but as soon as possible I think we should avoid having two implementations; but that goes without saying I guess.

Reply:
For sure, but flip-vector is already public and relied upon by res  (and other bits of ecl), and for starters it does no harm. I think we can afford to wait before we purge util, as long as we don't rely on it in e3.